### ISSUE TITLE
RBM doesn't learn with default step function

### ISSUE THREAD

**Issue by Treiblesschorle** [deeplearning4j/deeplearning4j#1188]
Im trying to learn a RBM with one softmax layer on MNIST dataset with the following config: https://gist.github.com/Treiblesschorle/73ef2d3e84ea59610bb8

Following this paper: http://yosinski.com/media/papers/Yosinski2012VisuallyDebuggingRestrictedBoltzmannMachine.pdf

the activations before training on the first batch look like this:
![activations before training](https://cloud.githubusercontent.com/assets/7195415/13316131/d8c0f676-dbae-11e5-82dd-dc26f9914b2e.PNG)

Next, I ran two experiments once not setting the step function and once using DefaultStepFunction.
If I do not set the step function in the configuration the RBM doesn't learn anything and the error rises. I ran an experiment with 10 epochs of pretraining. Visualization during training looks like this:
http://abload.de/img/high_error1dse3.png
and extracted filters for the first five neurons(bottom) and activations(top) after training:
http://abload.de/img/filters_hight_errorvhkme.png
After 10 epochs of pretraining the error on finetune stays on something like 2 and its not able to learn something.

The second experiment with exactly the same setup except that i set the step function to DefaultStepFunction looks like this during 10 epochs of pretraining:
http://abload.de/img/low_errorp6sjb.png
and filters and activation after training:
http://abload.de/img/filters_activation84so4.png

Everything looks like the things described in the Yosinski paper. Looking at the filters you can see that it learned some shapes of the MNIST numbers. However, after pretrainig the error on the softmax layer explodes. It's like 20000 after 10 epochs of finetuning. 
If I use the DefaultStepFunction it seems like it is maximizing the error on the softmax layer and minimizing it for the RBM. And exactly the opposite happens when I do not change the step function.

Has anyone an idea what could be the problem here?

**Comment by Cryddit** [comment]
I found this today.  If I'm reading the code right (and it's an unfamiliar language to me so I might not be) it's multiplying by a weight as it goes from the input layer to the reduced-dimensionality vector, and then multiplying by the same weight (rather than its reciprocal, which would be correct) when going from the reduced-dimensionality vector to the attempted reconstruction of input. 

Anyway, this makes the restricted Boltzmann Autoencoder explode.  If you don't put a hard regularizer like L2 clipping on it, the weights grow (or if they start  less than 1, shrink) exponentially.  And if you do, then they go to whatever limit the regularization allows and noodle around there without learning anything.  I was about to report this bug, but it looks like you beat me to it.

**Comment by Cryddit** [comment]
I was trying to figure out why several examples weren't working.  A comment says 

NOTE: This example has not been tuned.  It requires additional work to produce sensible results.  

Which is true, but the work it requires isn't in the same file as the example code.  Someone attempting to solve it there will be frustrated.

**Comment by LanceBarnett** [comment]
@Cryddit I thought the weights in an RBM were symmetric. Why should the reconstruction use reciprocal weights? Is there a source where you have seen that?

**Comment by Cryddit** [comment]
It's just math.  Consider a 3-into-2 dimensionality reduction.  Where A,B,C are the activations on one side, and D,E are the activations on the other side, the weights are u-z, and biases are a,b,c then you have

(A+a)u + (B+b)w + (C+c)y = D    &     (A+a)v + (B+b)x + (C+c)Z = E 

and depending on the reciprocal question you get either 

Du + Ev = A+a    &    Dw + Ex = B+b    &     Dy + Ez = C+c,  
or 
D/u + E/v = A+a    &    D/w + E/x = B+b    &    D/y + E/z = C+c.  

In any case where A,B,C are restricted to just four sets of values (so dimensionality reduction is possible at all) you should be able to find values of a,b,c,u,v,w,x,y,z that make these equations have a simultaneous solution.  With the second set you can, and with the first set you can't. Try a few cases.

**Comment by Cryddit** [comment]
Because it's tedious to do multivariate algebra by hand, here's a simple regression solver, in C.  This will home in on a set of weights that make the first set of equations true no matter what four sets of values you plug in for the contents of the arrays As, Bs, and Cs. 

You can't write a solver for the other set of equations because there's no general solution. 

`#include <stdio.h>


double absolute(const double arg){return (arg > 0 ? arg : -arg);}

int case1adjust(double *a, double *b, double *c, double *u, double *v, double *w, double *x, double *y, double *z,
		const double A, const double B, const double C){

    double D = (A + *a)* *u + (B + *b)* *w + (C+*c)* *y;
    double E = (A + *a)* *v + (B + *b)* *x + (C+*c)* *z;

    double Aprime = D/ *u + E/ *v - *a;
    double Bprime = D/ *w + E/ *x - *b;
    double Cprime = D/ *y + E/ *z - *c;

    double diffA = Aprime - A;
    double diffB = Bprime - B;
    double diffC = Cprime - C;

    *a += 0.001 * (diffA - diffB);
    *a += 0.001 * (diffA - diffC);
    *b += 0.001 * (diffB - diffA);
    *b += 0.001 * (diffB - diffC);
    *c += 0.001 * (diffC - diffA);
    *c += 0.001 * (diffC - diffB);

    *u += 0.001 * diffA;
    *v += 0.001 * diffA;
    *w += 0.001 * diffB;
    *x += 0.001 * diffB;
    *y += 0.001 * diffC;
    *z += 0.001 * diffC;
    return (absolute(diffA) + absolute(diffB) + absolute(diffC));
}

#define LOTS 100000

int main(){
    double a, b, c, u, v, w, x, y, z;
    double As[4] = { 1.0, 2.8, 3.2, 2.2 };
    double Bs[4] = { 2.8, 3.0, 1.9, 1.6 };
    double Cs[4] = { 1.4, 2.2, 9.1, 2.1 };

    a = b = c = u = v = w = x = y = z = 0.0;

    int count, iter;
    double totl = 1.0;
    for (iter = 0; iter < LOTS && totl > 0.05; iter++){
	totl = 0.0;
	for (count = 0; count < 4; count++)
	    totl += case1adjust(&a, &b, &c, &u, &v, &w, &x, &y, &z, As[count], Bs[count], Cs[count]);
    }
    if (iter < LOTS) printf("case 1 success!\n");
    else printf("failure!\n");

    a = b = c = u = v = w = x = y = z = 0.0;
    totl = 1.0;

    for (iter = 0; iter < LOTS && totl > 0.05; iter++){
	totl = 0.0;
	for (count = 0; count < 4; count++)
	    totl += case2adjust(&a, &b, &c, &u, &v, &w, &x, &y, &z, As[count], Bs[count], Cs[count]);
    }
    if (iter < LOTS) printf("case 2 success!\n");
    else printf("case 2 failure!\n");

}
`

**Comment by Cryddit** [comment]
Anyhow, the mathematical identities for reconstruction work in multiply/divide but don't work for multiply/multiply.  The code above is just a little empirical solver that demonstrates you can't find a  disproof-by-counterexample.  If you add more cases than four, you're no longer guaranteed that a set of weights allowing reconstruction exists with only two activations to encode them in, because more than four cases means you might not have a consistent encoding in only two dimensions. 

It's a little better than that, actually, because I allowed adjustable free variables for the biases; with the displacements I think it could solve for reconstruction of up to eight arbitrary cases.  But the proof of that isn't trivial, and I don't have it in my head.

**Comment by Cryddit** [comment]
Ergh.  I hate this.  I had a dumb mistake in the solver I just posted.  It should have been this instead.

```
#include <stdio.h>


double absolute(const double arg){return (arg > 0 ? arg : -arg);}

int case1adjust(double *a, double *b, double *c, double *u, double *v, double *w, double *x, double *y, double *z,
		const double A, const double B, const double C){

    double D = (A + *a)* *u + (B + *b)* *w + (C+*c)* *y;
    double E = (A + *a)* *v + (B + *b)* *x + (C+*c)* *z;

    double Aprime = D/ *u + E/ *v - *a;
    double Bprime = D/ *w + E/ *x - *b;
    double Cprime = D/ *y + E/ *z - *c;

    double diffA = Aprime - A;
    double diffB = Bprime - B;
    double diffC = Cprime - C;

    *a -= 0.001 * (diffA - diffB);
    *a -= 0.001 * (diffA - diffC);
    *b -= 0.001 * (diffB - diffA);
    *b -= 0.001 * (diffB - diffC);
    *c -= 0.001 * (diffC - diffA);
    *c -= 0.001 * (diffC - diffB);

    *u += 0.001 * diffA;
    *v += 0.001 * diffA;
    *w += 0.001 * diffB;
    *x += 0.001 * diffB;
    *y += 0.001 * diffC;
    *z += 0.001 * diffC;
    return (absolute(diffA) + absolute(diffB) + absolute(diffC));
}

#define LOTS 100000

int main(){
    double a, b, c, u, v, w, x, y, z;
    double As[4] = { 1.0, 2.8, 3.2, 2.2 };
    double Bs[4] = { 2.8, 3.0, 1.9, 1.6 };
    double Cs[4] = { 1.4, 2.2, 9.1, 2.1 };

    a = b = c = u = v = w = x = y = z = 0.1;

    int count, iter;
    double totl = 1.0;
    for (iter = 0; iter < LOTS && totl > 0.05; iter++){
	totl = 0.0;
	for (count = 0; count < 4; count++)
	    totl += case1adjust(&a, &b, &c, &u, &v, &w, &x, &y, &z, As[count], Bs[count], Cs[count]);
    }
    if (iter < LOTS) printf("case 1 success!\n");
    else printf("failure!\n");

}

```

The differences:  signs on the adjustments to the biases, and initializing the weights/biases to non-zero so as to avoid NaN results.

**Comment by agibsonccc** [comment]
This strikes me as more of a tuning question. Closing this.

**Comment by nyghtowl** [comment]
Using negative default step seemed to fix this at the time using rc3.10 (noted above) but this should be revisited for review.

**Comment by agibsonccc** [comment]
@nyghtowl how is this coming?

**Comment by nyghtowl** [comment]
This issue is resolved in this PR: https://github.com/deeplearning4j/deeplearning4j/pull/2194

Still working through some test failures (unrelated to step) before PR merge.
